core/user_controller/tests.py

```diff
@@ -0,0 +1,507 @@
+        course.save()
+        course_count += 1
+        for i in range(course_count):
+            instructor = create_user("Instructor", user_count)
+            user_count += 1
+            section = create_course_section(course.id, i, instructor)
+            ta = create_user("TA", user_count)
+            user_count += 1
+
+            create_ta_course_assignment(course, ta)
+
+            lab_section = create_lab_section(course, i)
+            if i % 2 == 1:
+                create_lab_assignment(lab_section, ta)
+
+class TestGetUser(TestCase):
+    def setUp(self):
+        self.course_list = [
+            ('Test1', 'Software Engineering'),
+            ('Test2', 'Software Development'),
+            ('Other3', 'Calculus 777'),
+            ('1000000000', '7777777777'),
+            ('CS101', 'Intro to CS')
+        ]
+        setup_database(self.course_list)
+        self.unassigned_user = User.objects.create_user(role='Admin', email='EMAIL_TEST', password='PASSWORD_TEST',
+                                                        first_name='AdminF_name', last_name='AdminL_name',
+                                                        username='AdminUsername')
+        self.unassigned_user.save()
+        self.course = Course.objects.get(course_code='CS101')
+
+    def setup_lab_assignments(self):
+        # Creating a Lab Section and assigning it to TA
+        self.semester = Semester.objects.create(semester_name='Fall 2023', start_date='2023-09-01',
+                                                end_date='2023-12-31')
+        self.instructor_user = User.objects.create_user(role='Instructor', email='INST_TEST', password='PASSWORD_TEST',
+                                                        first_name='InstructorF_name', last_name='InstructorL_name',
+                                                        username='InstructorUsername')
+        self.ta_user = User.objects.create_user(role='TA', email='TA_TEST', password='PASSWORD_TEST',
+                                                first_name='TAF_name', last_name='TAL_name', username='TAUsername')
+
+        self.course_section = CourseSection.objects.create(
+            course=self.course, instructor=self.instructor_user, course_section_number=1,
+            start_time='09:00', end_time='10:00'
+        )
+
+        lab_section = LabSection.objects.create(
+            course=self.course, lab_section_number=1, start_time='10:00', end_time='12:00'
+        )
+        TALabAssignment.objects.create(ta=self.ta_user, lab_section=lab_section)
+
+    def test_NoInput(self):
+        with self.assertRaises(ValueError):
+            UserController.getUser(None)
+
+    def test_InvalidInput(self):
+        with self.assertRaises(ValueError):
+            UserController.getUser("invalid_id")
+
+    def test_validInputs(self):
+        self.setup_lab_assignments()
+
+        for user in User.objects.all():
```

Comment:
this will need to be refactored for the data types. The data classes don't have good equality so you'll have to manually verify the data or add __eq__ method to the data classes (which would be somewhat long because list order doesn't matter)